spock-core/src/main/java/org/spockframework/runtime/SpockNode.java

```diff
@@ -126,4 +130,34 @@ protected static ExecutionMode toExecutionMode(org.spockframework.runtime.model.
       .map(er -> new ExclusiveResource(er.getKey(), toLockMode(er.getMode())))
       .collect(Collectors.toSet());
   }
+
+  @Override
+  public Set<TestTag> getTags() {
+    // Lazy initialisation
+    if (tags == null) {
+      tags = new HashSet<>();
+      SpecInfo specInfo;
+      // Node info can be either FeatureInfo or SpecInfo
+      if (getNodeInfo() instanceof FeatureInfo) {
+        FeatureInfo featureInfo = (FeatureInfo) getNodeInfo();
+        addTestTagsFrom(featureInfo.getTags());
+        specInfo = featureInfo.getSpec();
+      }
+      else {
+        specInfo = (SpecInfo) getNodeInfo();
+      }
+      while (specInfo != null) {
```

Comment:
🤔 thinking about it more, it would probably make more sense to delegate the aggregation to the extension completely, and just return the tags that are defined on the node. Meaning the extension will take care of adding the tag to all the necessary nodes.

Reply:
I disagree. The extension is already taking care of transforming the annotations to Spock tags. That is its purpose, its responsibility. I am also explaining that in the [Javadoc](https://github.com/kriegaex/spock/blob/10cd9e28afa229f58db5ee5b81a3250cda150be9/spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TagExtension.java#L23-L27):

> Processes `@Tag` annotations on specifications and features, converting them to Spock tags which can potentially be used by reporting extensions.

I.e., the extension as such is working completely independently of JUnit. I think, we should also keep it that way. The same `@Tag` annotations can be used for reporting purposes, e.g. for reporting test results or test runtimes per user-defined (possibly overlapping) categories. This would all work without ever activating tag filtering in JUnit at all. We should not let JUnit bleed in here.

Only in the next sentence I am explaining for the user's benefit one of the use cases for tags, which Spock itself currently supports:

> More importantly, it also makes the tag information available to JUnit 5 as platform tags in order to enable filtering, e.g. from build tools like Maven or Gradle.

But that functionality is handled where it belongs: in `SpockNode` classes which implement JUnit platform interfaces.

Reply:
As for why the node should in fact traverse the hierarchy, I explained that above already.

Concerning `isValid`, it would really cause `org.junit.platform.commons.PreconditionViolationException: Tag name [foo bar] must be syntactically valid` when using tags not conforming to the JUnit platform rules. That would be easy to mitigate by simply requiring the same from `@Tag` values, either explicitly when scanning them in the `TagExtension` or by simply documenting (with a link to the JUnit paragraph explaining the syntax) that we deliberately impose the same restriction, because one possible use case for those tags would be to use them for JUnit tag filtering. I am for the latter approach, not for enforcing it. The restriction is still so weak that it would not affect the general usage of `@Tag` - definitely not enough for me to justify to create an extra `ITaggable` and `@TestTag` just for JUnit and then something similar for general-purpose tags only for Spock. KISS! DRY! Avoid bloat.

Reply:
https://github.com/spockframework/spock/pull/1437/commits/d9c8c50e85f62a81e427a587030006cd13ac5e53 adds javadoc for the tag value.

Reply:
My tendency is to keep it like this and let Spock users figure out how to fix their tags, reading the JUnit exception and the Javadoc for our `Tag.value`. Of course, instead of

```java
    specElementTags.stream()
      .filter(tag -> tag.getKey().equals(TagExtension.TAG_EXTENSION_KEY))
      .map(Tag::getName)
      .map(TestTag::create)
      .forEach(tags::add);
```
we could use
```java
    specElementTags.stream()
      .filter(tag -> tag.getKey().equals(TagExtension.TAG_EXTENSION_KEY))
      .map(Tag::getName)
      .filter(TestTag::isValid)
      .map(TestTag::create)
      .forEach(tags::add);
```
But then there would not be any warning or error. Like you said before, users might wonder what happened to their tags. Another variant would be
```java
    specElementTags.stream()
      .filter(tag -> tag.getKey().equals(TagExtension.TAG_EXTENSION_KEY))
      .map(Tag::getName)
      .map(name -> {
        try {
          return Optional.of(TestTag.create(name));
        }
        catch (PreconditionViolationException e) {
          System.err.println("Ignoring invalid tag name '" + name + "'");
          return Optional.<TestTag>empty();
        }
      })
      .filter(Optional::isPresent)
      .map(Optional::get)
      .forEach(tags::add);
```
But I have noticed that Spock tries to avoid logging stuff to the console or to loggers. So that might not be the Spock way of doing it. It also would potentially log the same tag name many times, if e.g. a tag is defined on class level and the class has many methods, each one inheriting the tag and JUnit therefore trying to create it.

Another option would be to add a configurable option to the `TagExtension` where users can opt in to ignore invalid tags, while the default would be to pass them through to JUnit, causing exceptions. Then whether we apply `.filter(TestTag::isValid)` would depend on that flag.

The least attractive option for me would be to check all tags in the extension before transforming them to Spock tags. Why do JUnit's job there? Also, it would bleed JUnit logic into the extension code, which I want to avoid, Like I already explained in detail.

Reply:
OK, despite my uneasy feeling about pulling up any JUnit specifics into the `TagExtension`, I have added a naming syntax check using the JUnit's `TestTag::isValid` right into the annotation scanning process, throwing an `InvalidTagNameException` upon violation. See commit 93507247.

Back to the actual review topic here, I still think it would be move the aggregation logic into the extension. It ought to stay in the nodes. Are you able to accept that? I think I gave good reasons why that does not belong into the extension, because it is 100% JUnit-specific and JUnit is dealt with in the `*Nodes` classes.

Reply:
Please note that 886dad34 moves aggregation to `ITaggable` and its implementing classes. The commit comment explains in detail why I did that and what the benefits are. I.e., taggables now have their own notion of aggregation and it is no longer needed in either `SpockNode` or `TagExtension`. I am therefore resolving this conversation. Let us continue to discuss tag aggregation in the review of my new commit.